Explore: Add logs navigation to request more logs

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
This PR is part of [adding Logs Navigation to Explore](https://www.figma.com/file/7ZHtMoaK4RxBz3b6m6U0cW/Show-more-logs-%2B-Sorting?node-id=611%3A11713). 

Previous related PRs:
- https://github.com/grafana/grafana/pull/32978
- https://github.com/grafana/grafana/pull/33069

This PR adds Logs Navigation on the right side of listed logs. I'll work iteratively in next PRs on improving this. The goal is to make PRs small.

https://user-images.githubusercontent.com/30407135/115712357-f95baa00-a374-11eb-9f77-3a04b66ae5a2.mov
![image](https://user-images.githubusercontent.com/30407135/115709586-ab917280-a371-11eb-9d71-a1ca689edc8d.png)


**Things that this PR covers:**
- Adds the log navigation/pagination to the right that refreshes when new query is run
- Each page is showing logs range - from - to based on visible range. It is the same range as in histogram.
- Clicking at page currently runs the same query as was run for the chunk
- Adds a button in logs navigation to fetch new logs. Clicking on it runs a new query. While loading state, we see spinner and button and pages are disabled.
- Flipping order of logs flips the logs navigation as well, so it is in the same order

**Things to tackle in the next PRs:**
- Adding "Newer logs" button
- Caching of  results (logs) from previous queries
- Adjusting the length of page line to length of received range
- Remove scrolling of pagination and replace with something better
- General improvements - adding number of received logs for chunk,...
- Histogram

**Related issue:**
https://github.com/grafana/grafana/issues/19354

ivanahuckova committed:
Update layout for new logs navigation

ivanahuckova committed:
Add LogsNavigation component

ivanahuckova committed:
WIP clean up of css

ivanahuckova committed:
Add clear navigation and styling

ivanahuckova committed:
Add clearing to queyrRows run Query

ivanahuckova committed:
Make querying and displaying reliable

ivanahuckova committed:
Rename chunks to pages

ivanahuckova committed:
Refactor, remove logsNavigationCleared from navigation

ivanahuckova committed:
Rmove clear logs navigation

ivanahuckova committed:
Add flipping

ivanahuckova committed:
Update styling

ivanahuckova committed:
Add test

ivanahuckova committed:
Merge branch 'master' into ivana/explore-navigation

ivanahuckova committed:
Fix import

ivanahuckova committed:
Update Wrapper test

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRows.tsx:
```diff
@@ -24,9 +24,6 @@ export interface Props extends Themeable {
   previewLimit?: number;
-  // Passed to fix problems with inactive scrolling in Logs Panel
-  // Can be removed when we unify scrolling for Panel and Explore
-  disableCustomHorizontalScroll?: boolean;
```
We can finally get rid of this and we are using Custom Scrollbar in Logs component

ivanahuckova commented on public/app/features/explore/LogsNavigation.tsx:
```diff
@@ -0,0 +1,243 @@
+
+export default memo(LogsNavigation);
+
+const getStyles = stylesFactory((theme: GrafanaTheme, oldestLogsFirst: boolean, loading: boolean) => {
```
I'll work more on improving CSS in next iterations as I still need to add Newer button and Scroll to top button.

ivanahuckova commented on public/app/features/explore/Wrapper.test.tsx:
```diff
@@ -64,7 +64,7 @@ describe('Wrapper', () => {
     (datasources.loki.query as Mock).mockReturnValueOnce(makeLogsQueryResponse());
 
     // Make sure we render the logs panel
-    await screen.findByText(/^Logs$/i);
```
As we have in Older logs button the word logs, I needed to remove case insensivity to make it work

ivanahuckova commented on public/sass/pages/_explore.scss:
```diff
@@ -157,6 +157,7 @@
   flex: 1 1 auto;
   flex-direction: column;
   padding: $dashboard-padding;
+  padding-bottom: 0px;
```
```suggestion
```

ivanahuckova committed:
Update public/sass/pages/_explore.scss

Elfo404 commented on public/app/features/explore/LogsNavigation.tsx:
```diff
@@ -0,0 +1,243 @@
+              </div>
+            ))}
+          </div>
+          <div className={styles.filler}></div>
```
you should be able to achieve the same without creating an element by doing

```css

      &::after {
        content: '';
        display: block;
        background: repeating-linear-gradient(
          135deg,
          ${theme.colors.bg1},
          ${theme.colors.bg1} 5px,
          ${theme.colors.bg2} 5px,
          ${theme.colors.bg2} 15px
        );
        width: 3px;
        height: inherit;
        margin-bottom: 8px;
      }
```

in `pagesWrapper` instead of creating a new element

Elfo404 commented on public/app/features/explore/Logs.tsx:
```diff
@@ -460,5 +477,10 @@ const getStyles = stylesFactory((theme: GrafanaTheme) => {
+    logsSection: css`
+      display: flex;
+      flexdirection: row;
+      height: 95vh;
```
if we use `max-height` here and in https://github.com/grafana/grafana/pull/33259/files#diff-52f1275b53c883d4dcd6cfa6ab6c164a37f808e481f131abe880b4f09877b515R171 we could make it nicer in the situations you have only a few logs:

![Screenshot 2021-04-26 at 14 10 13](https://user-images.githubusercontent.com/1170767/116087785-2243ae00-a699-11eb-9a77-c7266642d34e.png)

instead of 
![Screenshot 2021-04-26 at 14 11 09](https://user-images.githubusercontent.com/1170767/116087926-430c0380-a699-11eb-87fa-8df51bcc2e6e.png)

Elfo404 reviewed (commented):
I have a few more questions/comments, leaving some now and I'll come back to this later in the afternoon

btw, it looks great! 🚀

ivanahuckova committed:
Update public/app/features/explore/Logs.tsx

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

Elfo404 commented on public/app/features/explore/LogsNavigation.tsx:
```diff
@@ -0,0 +1,243 @@
+          }}
+          disabled={loading}
+        >
+          <div className={styles.navButtonContent}>
+            {loading ? <Spinner /> : <Icon name="angle-up" size="lg" />}
+            <div>Older</div>
+            <div>logs</div>
+          </div>
```
if you set
```
      white-space: normal;
```

in `navButtonContent` you might be able to not wrap each child in a div

ivanahuckova replied:
👍 🚀

ivanahuckova committed:
Add spinner to page to prevent jumpy time

ivanahuckova commented:
Added a loading state for pages as well - to prevent jumpy time (as we don't know what the final range is until we receive logs). This also provides more clarity on what is happening. 
https://user-images.githubusercontent.com/30407135/116119099-567d9580-a6be-11eb-847c-49e04317e65b.mov

ivanahuckova committed:
Create standard size for page

ivanahuckova committed:
Fix test

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/explore-navigation

aocenas commented on public/app/features/explore/LogsNavigation.tsx:
```diff
@@ -0,0 +1,256 @@
+  useEffect(() => {
+    const initialRange = absoluteRange.to - absoluteRange.from;
+    setRequestRange(initialRange);
+    // We want to set requestRange only when the first query is run
```
We don't really need effect for this or state. We can just do:
```
rangeRef = useRef()
function getRange() {
  rangeRef.current || rangeRef.current = absoluteRange.to - absoluteRange.from
}
```

aocenas commented:
Also some issues:
1. The timeline shows after all the logs not next to it.
![Screenshot from 2021-04-30 12-44-08](https://user-images.githubusercontent.com/1014802/116685791-34805d80-a9b3-11eb-948c-3e7dba0e132f.png)

2. Not issue with this PR but weirdly it seems like loki actually does not stick to the time range. If I query last 5min of logs I always get 1000 logs specified by the limit even if that goes over the 5min range. @slim-bean is this a known bug/feature?

3. In case there are no more logs the timerange is set to 0 width for some reason. I would expect some set interval so if there is a gap I can actually reach the logs after the gap.
![Screenshot from 2021-04-30 13-00-19](https://user-images.githubusercontent.com/1014802/116686402-08b1a780-a9b4-11eb-8e34-2b9992a09f7e.png)

aocenas commented:
@slim-bean 
> Not issue with this PR but weirdly it seems like loki actually does not stick to the time range. If I query last 5min of logs I always get 1000 logs specified by the limit even if that goes over the 5min range. @slim-bean is this a known bug/feature?

Seems like I had some old image where this was happening in new one this seems ok.

ivanahuckova committed:
Fix postioning and useRefs

ivanahuckova commented on public/app/features/explore/LogsNavigation.tsx:
```diff
@@ -0,0 +1,256 @@
+      const index = newPages.findIndex((page) => page.queryRange.to === absoluteRange.to);
+      setCurrentPageIndex(index);
+    }
+    // We don't want to add expectedRange as we want to run this only is absolute and visible range changes
```
Hmmm, good point, I was thinking about this more and decided to use refs for rangeSpan, expectedRange and also expectedQueries as we don't want to re-render LogsNavigation when these change. We just want to keep track of these variables. This way, I could use all state and props in dependency array.

ivanahuckova commented:
>The timeline shows after all the logs not next to it. 
Fixed. I incorrectly resolved this merge conflict, but it is fixed now. 

>In case there are no more logs the timerange is set to 0 width for some reason. I would expect some set interval so if there is a gap I can actually reach the logs after the gap.

Not sure what do you mean, the timeRange is set to 1h (11:38 - 10:38) (which I think was your initial time range)

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/explore-navigation

aocenas commented:
> Not sure what do you mean, the timeRange is set to 1h (11:38 - 10:38) (which I think was your initial time range)

Ach ok my bad I did not notice the hour changed. My time range was actually 5 or 10min that's why I did not expect there to be exact hour difference. Is the one hour some default or should that be something different if I had short initial time range?

ivanahuckova commented:
>Ach ok my bad I did not notice the hour changed. My time range was actually 5 or 10min that's why I did not expect there to be exact hour difference. Is the one hour some default or should that be something different if I had short initial time range?

It was set based on first query (which was bug) - I changed it so it is set when we add a first page.

aocenas reviewed (approved):
Looks good have just 2 nitpicks questions that don't have to be addressed here:
1. I wonder if we could at some point remove the custom react scroll bars? They seem to have some weird performance issues that create and initial lag when dragging the scrollbar.
2. Sort of feel like it's not needed to remove the time information when loading. Seems to me like it would mean the time range will change but that does not happen right? Not a big deal though and don't have a good idea where else to put it either.

https://user-images.githubusercontent.com/1014802/116894338-3b68d380-ac32-11eb-9b61-ad2acef4c0f2.mp4

Elfo404 reviewed (approved):
not sure if you accidentally missed `max-height: 95vh` in https://github.com/grafana/grafana/pull/33259/files#diff-6d428f1742c9ebba65767b8d6ea3e6fc55c24978f60dcc96f1dbb782ccd64c2dR413 or it was intentional (per https://github.com/grafana/grafana/pull/33259/files#r620281674)

but looks and works like a charm! 🚀

ivanahuckova commented:
>not sure if you accidentally missed max-height: 95vh

Fixing it. I swear I added it and it disappeared on its own. 👻  

>I wonder if we could at some point remove the custom react scroll bars? They seem to have some weird performance issues that create and initial lag when dragging the scrollbar.

Torkel mentioned that on Linux without CustomScrollbar it looks weird. But in this feature there is definitely a plan - I have added in the description the - next steps - and one of them is **Remove scrolling of pagination and replace with something better**

>Sort of feel like it's not needed to remove the time information when loading. Seems to me like it would mean the time range will change but that does not happen right? Not a big deal though and don't have a good idea where else to put it either.

When we run the query, we want to keep spinner as the time might change and also it shows user that the next logs are being fetched. For the cached logs (when we add caching), I agree that we don't need to do that.

ivanahuckova committed:
Add maxHeight

baurine commented:
hi, thanks for implementing this feature, when can we use it?

ivanahuckova commented:
Hello @baurine It will be part of the next release. The beta should be out in ~1 week.
